### PR TITLE
Convert each plugin to a function from a class to an overriding class

### PR DESCRIPTION
**Pull request by andy-ms**
| Q                 | A
| ----------------- | ---
| Bug fix?          | no
| Breaking change?  | no
| New feature?      | no
| Deprecations?     | no
| Spec compliancy?  | no
| Tests added/pass? | tests pass
| Fixed tickets     | 
| License           | MIT

Instead of assigning functions to a parser instance, this creates a class with overridden methods.
The class is a function of the plugins (only "jsx", "flow", and "estree"), and this function is cached.

Based on my earlier comment: https://github.com/babel/babylon/pull/393#issuecomment-292623114
Like #393, this avoids fixing the indent inside each plugin, in order to have a smaller diff.

This will improve performance because we aren't assigning functions directly to the instance any more.
I ran the [perf script](https://gist.github.com/andy-ms/138e270e510620945df1499181c05892) again on my home machine just to be sure:

| branch | time |
| --- | --- |
| `master` | 26.8 26.8 27.1 26.0 26.0 |
| `plugin_class` | 12.8 13.0 13.1 13.3 12.7 |

### CODE REVIEW

**Comment by hzoo** [comment]
Hey @andy-ms! It looks like one or more of your builds have failed. I've copied the relevant info below to save you some time.

* [Node vlts/* - linux](https://travis-ci.org/babel/babylon/jobs/220314589)
* [Node vlts/* - linux](https://travis-ci.org/babel/babylon/jobs/220314590)

**Comment by DanielRosenwasser** [comment]
Hey @loganfsmyth @hzoo, looks like the tests are fixed. Can you take another look?

**Comment by STRML** [comment]
> Like #393, this avoids fixing the indent inside each plugin, in order to have a smaller diff.

I so wish this were a default, but you can view the diff via https://github.com/babel/babylon/pull/459/files?w=1 (note the `?w=1`) to avoid highlighting whitespace diffs.

**Comment by andy-ms** [comment]
@STRML Thanks!

File: src/index.js [hunk 1]
```diff
@@ -32,3 +32,39 @@ export function parseExpression(input, options) {
...
+  const key = pluginList.join("/");
+  let cls = parserClassCache[key];
+  if (!cls) {
+    cls = Parser;
+    for (const plugin of pluginList) {
+      cls = plugins[plugin](cls);
+    }
+    parserClassCache[key] = cls;
+  }
+  return  cls;
```
Comment (hzoo) [inline src/index.js]: `return  cls;` guess we didn't have a rule for this, I guess we should land that prettier pr later

**Comment by danez** [comment]
I ran my [perf script](https://github.com/danez/babylon_performance) and the results are impressive. (Up to 50% faster)

Before
```
name                            run   
./fixtures/angular.js           502ms 
./fixtures/backbone.js          60ms  
./fixtures/ember.debug.js       1468ms
./fixtures/jquery.js            279ms 
./fixtures/react-with-addons.js 576ms 
```

after
```
name                            run  
./fixtures/angular.js           298ms
./fixtures/backbone.js          27ms 
./fixtures/ember.debug.js       779ms
./fixtures/jquery.js            141ms
./fixtures/react-with-addons.js 314ms
```

Thanks a lot

**Pull request merged**